x/stake/types/pool.go

```diff
@@ -52,108 +42,77 @@ func InitialPool() Pool {
 //____________________________________________________________________
 
 // Sum total of all staking tokens in the pool
-func (p Pool) TokenSupply() int64 {
-	return p.LooseTokens + p.UnbondedTokens + p.UnbondingTokens + p.BondedTokens
+func (p Pool) TokenSupply() sdk.Rat {
+	return p.LooseTokens.Add(p.BondedTokens)
 }
 
 //____________________________________________________________________
 
 // get the bond ratio of the global state
 func (p Pool) BondedRatio() sdk.Rat {
-	if p.TokenSupply() > 0 {
-		return sdk.NewRat(p.BondedTokens, p.TokenSupply())
+	if p.TokenSupply().GT(sdk.ZeroRat()) {
+		return p.BondedTokens.Quo(p.TokenSupply())
 	}
 	return sdk.ZeroRat()
 }
 
-// get the exchange rate of bonded token per issued share
-func (p Pool) BondedShareExRate() sdk.Rat {
-	if p.BondedShares.IsZero() {
-		return sdk.OneRat()
-	}
-	return sdk.NewRat(p.BondedTokens).Quo(p.BondedShares)
-}
+//_______________________________________________________________________
 
-// get the exchange rate of unbonding tokens held in validators per issued share
-func (p Pool) UnbondingShareExRate() sdk.Rat {
-	if p.UnbondingShares.IsZero() {
-		return sdk.OneRat()
+func (p Pool) looseTokensToBonded(bondedTokens sdk.Rat) Pool {
+	p.BondedTokens = p.BondedTokens.Add(bondedTokens)
+	p.LooseTokens = p.LooseTokens.Sub(bondedTokens)
+	if p.LooseTokens.LT(sdk.ZeroRat()) {
+		panic(fmt.Sprintf("sanity check: loose tokens negative, pool: %v", p))
 	}
-	return sdk.NewRat(p.UnbondingTokens).Quo(p.UnbondingShares)
+	return p
 }
 
-// get the exchange rate of unbonded tokens held in validators per issued share
-func (p Pool) UnbondedShareExRate() sdk.Rat {
-	if p.UnbondedShares.IsZero() {
-		return sdk.OneRat()
+func (p Pool) bondedTokensToLoose(bondedTokens sdk.Rat) Pool {
+	p.BondedTokens = p.BondedTokens.Sub(bondedTokens)
+	p.LooseTokens = p.LooseTokens.Add(bondedTokens)
+	if p.BondedTokens.LT(sdk.ZeroRat()) {
+		panic(fmt.Sprintf("sanity check: bonded tokens negative, pool: %v", p))
```

Comment:
Maybe we should have a `sdk.URat` type (not for this PR)

Reply:
yeah could maybe be good